### PR TITLE
Add more information about the current configuration in verbose mode

### LINKED ISSUE #603: exporting configuration
hello,
to help to create plugins or to debug some strange case it could be usefull to have the possibility to export the configuration.

something like --export-config / -ec that will show the config

extra : 
having multiple configuration format available like json/init or simple human readable

### PR DESCRIPTION
**Pull request by jubianchi**
This will fix #603

While working on this, we will check if CLI arguments correctly override the configration files (see #243)

### CODE REVIEW

**Comment by jubianchi** [comment]
We now have two things:

* CLI arguments correctly override configuration file (when possible)
  * `-g`/`--glob` cannot override the configuration file because its effects are immediatly applied to the runner
  * `-af`/`--autoloader-file` cannot override the configuration file because its effects are immediatly applied to the runner
  * `-bf`/`--bootstrap-file` cannot override the configuration file because its effects are immediatly applied to the runner
* Verbose mode displays a lot more informations:

```
Using ++verbose -g ./test*.php -ncc -xc IDEKEY=foobar arguments…
Using '/Users/julien.bianchi/repositories/atoum/atoum/.atoum.php' configuration file…
Using '/tmp/6e2deb60e60a380bb8dbe7b34cd86960.atoum.cache' autoloader cache file…
Using 'mageekguy\atoum\reports\realtime\santa' report…
Using 'IDEKEY=foobar' xDebug configuration…
Accepting 'php' file extensions…
Using './test.php' test file…
Using './test2.php' test file…
Code coverage is disabled…
```

any reviewer ?

**Comment by Hywan** [comment]
Looking.

File: classes/scripts/runner.php [hunk 1]
```diff
@@ -451,7 +493,7 @@ public function addTestsFromDirectories(array $directories)
 
 	public function addTestsFromPattern($pattern)
 	{
-		$this->runner->addTestsFromPattern($pattern);
+			$this->runner->addTestsFromPattern($pattern);
```
Comment (Hywan) [inline classes/scripts/runner.php]: Not sure about this one.

File: classes/scripts/runner.php [hunk 2]
```diff
@@ -1208,11 +1264,87 @@ protected function doRun()
...
+				{
+					$this->verbose($this->locale->_('Running \'%s::%s\' test method…', $testClass, $testMethod), 3);
+				}
+			}
+
+			$maxChildrenNumber = $this->runner->getMaxChildrenNumber();
+
+			if ($maxChildrenNumber !== null)
+			{
+				$this->verbose(sprintf($this->locale->__('Limitintg children number to \'%d\' process…', 'Limitintg children number to \'%d\' processes…', $maxChildrenNumber), $maxChildrenNumber), 2);
```
Comment (Hywan) [inline classes/scripts/runner.php]: Limiting*

File: resources/configurations/runner/santa.php.dist [hunk 3]
```diff
@@ -18,4 +18,4 @@ Generate a Santa report.
 $santaReport = new atoum\reports\realtime\santa();
 $santaReport->addWriter($stdOutWriter);
 
-$runner->addReport($cliReport);
+$runner->addReport($santaReport);
```
Comment (Hywan) [inline resources/configurations/runner/santa.php.dist]: lulz

**Comment by Hywan** [comment]
@jubianchi: The workd “information“ in English is “uncountable”, there is no plural form. So it's many and several of hundreds of information (no “s”).

**Comment by Hywan** [comment]
Instead of having `Using …` it would be better to reverse the display:
```
XDebug: …
Autoloader: …
```
Better than sentence I guess.

**Comment by Grummfy** [comment]
thanks. But for the export of configuration I was also thinking extracting if we enable options or set stuff like : loop, debug, path to test directory, namespace of test,... the idea back #603 is to parse the solution, and get information on atoum for the IDE, or the debuging.

Honestly I can't review the code because I'm not advanced enough in the core of atoum. sorry :(

**Comment by jubianchi** [comment]
@Hywan I edited the code so the output looks like this:

```
CLI arguments: ++verbose -g ./test*.php -xc IDEKEY=foobar -utr -ulr -ns foo bar\foo -mcn 3 -ebpc -nccfns foo -nccfm foo::bar -nccfc foo\bar\baz
Configuration file:
=> /Users/julien.bianchi/repositories/atoum/atoum/.atoum.php
Using no autoloader file…
Using no bootstrap file…
Autoloader cache file:
=> /var/folders/4g/16c91c3d507_645p1zszf4rc0000gp/T/6e2deb60e60a380bb8dbe7b34cd86960.atoum.cache
Reports:
=> mageekguy\atoum\reports\realtime\tap
xDebug configuration: IDEKEY=foobar
Accepted file extension: php
Test files:
=> ./test.php
=> ./test2.php
Accepted namespaces:
=> foo
=> bar\foo
Max children processes: 3
Code coverage:
=> Lines coverage
=> Branches coverage
=> Paths coverage
=> Excluded namespace:
==> foo
=> Excluded class:
==> foo\bar\baz
=> Excluded method:
==> foo::bar
```

@Grummfy I understand. This is the first step on something to quickly expose the configuration. We'll see if we can do anything better later, because, honestly, I don't think creating an exportable file would help ore than a clear output that user can copy/paste anywhere.

**Comment by Hywan** [comment]
@jubianchi: If you have one autoload cache, why the newline and the `=>`? Same for reports.

**Comment by jubianchi** [comment]
@Hywan In my configuration, I have only one autoloader and one report but we can add more ;)

**Comment by jubianchi** [comment]
Here is the final output for the verbose mode:

| Property                 | Value                                                                                                                      |
|--------------------------|----------------------------------------------------------------------------------------------------------------------------|
| Script class             | mageekguy\atoum\scripts\coverage                                                                                           |
| CLI arguments            | -f /Users/julien.bianchi/repositories/atoum/atoum-config-extension/tests/units/classes/extension.php +verbose -o /dev/null |
| Autoloader cache file    | /var/folders/4g/16c91c3d507_645p1zszf4rc0000gp/T/6e2deb60e60a380bb8dbe7b34cd86960.atoum.cache                              |
| Configuration file       | N/A                                                                                                                        |
| Accepted tags            | N/A                                                                                                                        |
| Accepted namespaces      | N/A                                                                                                                        |
| Accepted methods         |                                                                                                                            |
| Runner class             | mageekguy\atoum\runner                                                                                                     |
| Autoloader file          | N/A                                                                                                                        |
| Bootstrap file           | N/A                                                                                                                        |
| Max children processes   | N/A                                                                                                                        |
| Reports                  | mageekguy\atoum\reports\realtime\cli                                                                                       |
|                          | mageekguy\atoum\reports\asynchronous\clover                                                                                |
| xDebug configuration     | N/A                                                                                                                        |
| Accepted file extensions | php                                                                                                                        |
| Test files               | 1                                                                                                                          |
| Code coverage            | Lines coverage                                                                                                             |
| CC excluded namespaces   | N/A                                                                                                                        |
| CC excluded classes      | N/A                                                                                                                        |
| CC excluded methods      | N/A                                                                                                                        |

As you can see, the output is a perfectly valid markdown table. Users will be able to copy/paster this directly in issues when they have problems.

**Comment by Grummfy** [comment]
I test it, but is there a way to print the configuration but without executing the test?

**Comment by Grummfy** [comment]
and one of the missing information is the namespace mask used for the unit test

**Comment by jubianchi** [comment]
@Grummfy the namespace mask is not really something we can configure: we can't change it through a configuration file or using a CLI option.

Printing this information is hard right now as we would have to create an instance of the base test class and this class is abstract.

We can't extract this using reflection either because we would have to look for a constant, an annotation somewhere, a method implementation, ... too much things to lookup.

Before showing this we would have to refactor how atoum finds and uses this namespace pattern.

File: classes/iterators/filters/recursives/extension.php [hunk 4]
```diff
@@ -36,11 +36,9 @@ public function getAcceptedExtensions()
 
 	public function accept()
 	{
```
Comment (Grummfy) [inline classes/iterators/filters/recursives/extension.php]: change have no real link to the PR, but OK

File: classes/script/visitor.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace mageekguy\atoum\script;
+
+use
+	mageekguy\atoum\script,
+	mageekguy\atoum\scripts
+;
+
+interface visitor
```
Comment (Grummfy) [inline classes/script/visitor.php]: should we extends the previous existing visitor interface?

File: classes/scripts/runner.php [hunk 6]
```diff
@@ -809,10 +861,13 @@ function($script, $argument, $empty) {
 					function($script, $argument, $directories) {
 						if (sizeof($directories) <= 0)
 						{
-							throw new exceptions\logic\invalidArgument(sprintf($script->getLocale()->_('Bad usage of %s, do php %s --help for more informations'), $argument, $script->getName()));
+							throw new exceptions\logic\invalidArgument(sprintf($script->getLocale()->_('Bad usage of %s, do php %s --help for more information'), $argument, $script->getName()));
 						}
 
-						$script->excludeDirectoriesFromCoverage($directories);
+						$script
+							->resetExcludedDirectoriesFromCoverage()
```
Comment (Grummfy) [inline classes/scripts/runner.php]: je comprends pas trop le rapport avec la PR

File: classes/scripts/runner.php [hunk 7]
```diff
@@ -1170,48 +1239,14 @@ protected function doRun()
...
+			$table = new cli\table(array('Property', 'Value'));
+			$scriptVisitor = new atoum\script\visitors\verbose($table);
+			$runnerVisitor = new atoum\runner\visitors\verbose($this->getVerbosityLevel(), $table);
 
-			if ($autoloaderFile !== null)
-			{
-				$this->verbose($this->locale->_('Using \'%s\' autoloader file…', $autoloaderFile));
-			}
+			$scriptVisitor->visitRunner($this);
+			$runnerVisitor->visit($this->getRunner());
```
Comment (Grummfy) [inline classes/scripts/runner.php]: nothing for the extensions?

File: scripts/coverage.php [hunk 8]
```diff
@@ -1,34 +1,17 @@
 <?php
```
Comment (Grummfy) [inline scripts/coverage.php]: don't see why this change

File: tests/units/classes/iterators/filters/recursives/extension.php [hunk 9]
```diff
@@ -40,16 +40,41 @@ public function test__construct()
 	public function testAccept()
```
Comment (Grummfy) [inline tests/units/classes/iterators/filters/recursives/extension.php]: idem, je voix pas trop le rapport ...

par contre pas de unit test pour la sortie?

File: classes/cli/table.php [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+		$this->header = array_values($header);
+
+		return $this->computeRowMaxLength($this->header)->computeColumnWidths($this->header);
+	}
+
+	public function addRow(array $row)
+	{
+		$row = array_values($row);
+
+		for ($index = 1, $size = sizeof($row); $index <= $size; $index++)
```
Comment (Grummfy) [inline classes/cli/table.php]: we should start $index at 0, and "$index < $size" should become "$index < ($size - 1)", no?

File: classes/cli/table.php [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+				}
+
+				if (sizeof($row[$index - 1]) === 1)
+				{
+					$row[$index - 1] = array_pop($row[$index - 1]);
+				}
+			}
+		}
+
+		if (is_array($row[sizeof($row) - 1]) === false)
```
Comment (Grummfy) [inline classes/cli/table.php]: sizeof($row) => $size

File: classes/cli/table.php [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+			$lastColumn = array_pop($row);
+			$this->computeColumnWidths($this->rows[] = array_merge($row, array(array_shift($lastColumn))));
+
+			foreach ($lastColumn as $column)
+			{
+				$this->computeColumnWidths($this->rows[] = array_merge(array_fill(0, sizeof($row), ''), array($column)));
+			}
+		}
+
+		return $this->computeRowMaxLength($row);
```
Comment (Grummfy) [inline classes/cli/table.php]: honestly I will do all this merge and row mix a the end, no?

File: classes/scripts/runner.php [hunk 13]
```diff
@@ -1170,48 +1239,14 @@ protected function doRun()
...
-			if (sizeof($this->configFiles) > 0)
-			{
-				foreach ($this->configFiles as $configFile)
-				{
-					$this->verbose($this->locale->_('Using \'%s\' configuration file…', $configFile));
-				}
-			}
-
-			$autoloaderFile = $this->runner->getAutoloaderFile();
+			$table = new cli\table(array('Property', 'Value'));
```
Comment (Grummfy) [inline classes/scripts/runner.php]: why not a big if is verbose? or following if it's verbose or not, the visitor do something or nothing (dev/null mode)

Comment (Grummfy) [inline classes/scripts/runner.php]: use $this->verbosityLevel ?

Comment (Grummfy) [inline classes/scripts/runner.php]: perhaps we can have an interface for that part, with no default implementation (why a default one ?)

and then like the rest create a setVerboseRenderer(verboseRendererInterface $vr)? with a default renderer == table ?

File: classes/runner/visitors/verbose.php [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+	public function __toString()
+	{
+		return (string) $this->table;
+	}
+
+	public function visit(runner $runner)
+	{
+		$this->table
+			->addRow(array('Runner class', get_class($runner)))
```
Comment (agallou) [inline classes/runner/visitors/verbose.php]: table utilisé dans le visiteur. ça ne lie pas le visiteur avec la vue ? ne faudrait t-il pas passer par un objet contenant ces infos ?

File: classes/script/visitor.php (hunk 5 above)
Comment (jubianchi) [inline classes/script/visitor.php]: which one ?

File: classes/scripts/runner.php (hunk 6 above)
Comment (jubianchi) [inline classes/scripts/runner.php]: see #243

File: tests/units/classes/iterators/filters/recursives/extension.php (hunk 9 above)
Comment (jubianchi) [inline tests/units/classes/iterators/filters/recursives/extension.php]: see #243

File: classes/scripts/runner.php [hunk 15]
```diff
@@ -1170,48 +1239,14 @@ protected function doRun()
...
+			$scriptVisitor = new atoum\script\visitors\verbose($table);
+			$runnerVisitor = new atoum\runner\visitors\verbose($this->getVerbosityLevel(), $table);
```
Comment (agallou) [inline classes/scripts/runner.php]: utiliser le même constructeur pour les visitors ? (qui serait dans l'interface)

File: classes/script/visitor.php (hunk 5 above)
Comment (Grummfy) [inline classes/script/visitor.php]: the other visitor class , but in fact no? just same name but differnt purpose ;)

File: classes/cli/table.php (hunk 12 above)
Comment (jubianchi) [inline classes/cli/table.php]: if we do this just before the table rendering we have to do one more complete iteration over the whole table. Doing it here lets us compute values each time a row is added (so no global iteration on the whole table) and have those values directly available when rendering.

**Comment by jubianchi** [comment]
@Grummfy I'm not satisfied with the current implementation. It's poorly designed.

I'd like to close this issue so that we can start on something cleaner.
